### PR TITLE
Add support for runtimes flag

### PR DESCRIPTION
**Pull request by loosebazooka**
- move explode-war task into model space
- project -> model space data sharing via extensions

### CODE REVIEW

**Comment by loosebazooka** [comment]
@meltsufin

File: src/main/java/com/google/cloud/tools/gradle/appengine/FlexibleDataExtension.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package com.google.cloud.tools.gradle.appengine;
+
+import org.gradle.api.JavaVersion;
+
+import java.io.File;
+
+/**
+ * Created by appu on 7/11/16.
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/gradle/appengine/FlexibleDataExtension.java]: who is appu?

File: src/main/java/com/google/cloud/tools/gradle/appengine/StandardDataExtension.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * limitations under the License.
+ *
+ */
+
+package com.google.cloud.tools.gradle.appengine;
+
+import org.gradle.api.JavaVersion;
+
+/**
+ * Created by appu on 7/11/16.
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/gradle/appengine/StandardDataExtension.java]: ...

File: src/main/java/com/google/cloud/tools/gradle/appengine/FlexibleDataExtension.java (hunk 1 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/gradle/appengine/FlexibleDataExtension.java]: I can only assume some great guy

File: src/main/java/com/google/cloud/tools/gradle/appengine/AppEngineStandardPlugin.java [hunk 3]
```diff
@@ -67,29 +70,24 @@
...
-          }
-        });
+  private void configureTargetCompatibility(final Project project,
+      final StandardDataExtension projectData) {
+    project.afterEvaluate(new Action<Project>(){
+      @Override
+      public void execute(Project project){
+        JavaPluginConvention javaConvention = project.getConvention()
+            .getPlugin(JavaPluginConvention.class);
+        JavaVersion javaVersion = javaConvention.getTargetCompatibility();
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/gradle/appengine/AppEngineStandardPlugin.java]: Can getTargetCompatibility() return null, if targetCompatibility is not specified in build.gradle?

Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/gradle/appengine/AppEngineStandardPlugin.java]: I don't think so, lemme confirm

**Comment by meltsufin** [comment]
Other than the comment about targetCompatiblity, LGTM.

**Comment by loosebazooka** [comment]
@meltsufin added this hack (take a look, had to go home) I need to file bugs to make sure we propagate these up to appcfg...

File: src/main/java/com/google/cloud/tools/gradle/appengine/util/AppEngineWebXml.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ */
+public class AppEngineWebXml {
+
+  private final Document parsed;
+
+  private AppEngineWebXml(File appengineWebXml) {
+    try {
+      parsed = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(appengineWebXml);
+    } catch (SAXException | IOException | ParserConfigurationException e) {
+      throw new GradleException("Failed to parse appengine-web.xml");
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/gradle/appengine/util/AppEngineWebXml.java]: This is swallowing the causing exception. It needs to be brought up to the user's attention somehow.

File: src/main/java/com/google/cloud/tools/gradle/appengine/util/AppEngineWebXml.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    return new AppEngineWebXml(appengineWebXml);
+  }
+
+  public boolean isVm() {
+    try {
+      XPath xpath = XPathFactory.newInstance().newXPath();
+      String expression = "/appengine-web-app/vm/text()='true'";
+      return (Boolean) xpath.evaluate(expression, parsed, XPathConstants.BOOLEAN);
+    } catch (XPathExpressionException e) {
+      throw new GradleException("XPath evaluation failed on appengine-web.xml");
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/gradle/appengine/util/AppEngineWebXml.java]: Ditto regarding the cause.

File: src/main/java/com/google/cloud/tools/gradle/appengine/AppEngineStandardPlugin.java (hunk 3 above)
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/gradle/appengine/AppEngineStandardPlugin.java]: Were you able to confirm?

Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/gradle/appengine/AppEngineStandardPlugin.java]: Gradle defaults to the version of java that the user is building with.

File: src/main/java/com/google/cloud/tools/gradle/appengine/util/AppEngineWebXml.java (hunk 4 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/gradle/appengine/util/AppEngineWebXml.java]: yeah, I'll propagate.

File: src/test/java/com/google/cloud/tools/gradle/appengine/util/AppEngineWebXmlTest.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  @Test
+  public void testIsVm_true() throws IOException {
+    Assert.assertTrue(AppEngineWebXml.parse(createAppEngineWebXml("<vm>true</vm>")).isVm());
+  }
+
+  @Test
+  public void testIsVm_false() throws IOException {
+    Assert.assertFalse(AppEngineWebXml.parse(createAppEngineWebXml("<vm>false</vm>")).isVm());
+    Assert.assertFalse(AppEngineWebXml.parse(createAppEngineWebXml("<vm>TRUE</vm>")).isVm());
+    Assert.assertFalse(AppEngineWebXml.parse(createAppEngineWebXml("<vm>potato</vm>")).isVm());
```
Comment (meltsufin) [inline src/test/java/com/google/cloud/tools/gradle/appengine/util/AppEngineWebXmlTest.java]: maybe something other than :sweet_potato: ? 'NONE'?

Comment (loosebazooka) [inline src/test/java/com/google/cloud/tools/gradle/appengine/util/AppEngineWebXmlTest.java]: done.

Comment (meltsufin) [inline src/test/java/com/google/cloud/tools/gradle/appengine/util/AppEngineWebXmlTest.java]: I meant instead of :sweet_potato: - :tomato:  :)

Comment (loosebazooka) [inline src/test/java/com/google/cloud/tools/gradle/appengine/util/AppEngineWebXmlTest.java]: done done

**Comment by meltsufin** [comment]
nice! :ship:

**Pull request merged**